File: 백준/Silver/1966. 프린터 큐/프린터 큐.py
import sys
from collections import deque
case = int(sys.stdin.readline())

for _ in range(case):
    q = deque()
    n, m = list(map(int, sys.stdin.readline().split()))
    doc = list(map(int, sys.stdin.readline().split()))
    for i, d in enumerate(doc):
        q.append((d, i))
    
    cnt = 0
    while q:
        priority = max(q)[0]
        d = q.popleft()
        
        if d[0] != priority:
            q.append(d)
            
        else:
            cnt += 1
            if d[1] == m:
                break
        
    print(cnt)

File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());

        while (n-- > 0) {
            String[] testCase = br.readLine().split(" ");
            int m = Integer.parseInt(testCase[0]);
            int target = Integer.parseInt(testCase[1]);

            Deque<Documents> q = new ArrayDeque<>();
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int i = 0; i < m; i++) {
                int priority = Integer.parseInt(st.nextToken());
                q.addLast(new Documents(priority, i));
            }

           int count = 0;
           while (!q.isEmpty()) {
               int maxPriority = q.stream().mapToInt(doc -> doc.priority).max().orElse(1);
               Documents curDoc = q.removeFirst();

               if (curDoc.priority == maxPriority) {
                   count++;
                   if (curDoc.idx == target) {
                       sb.append(count).append("\n");
                       break;
                   }
               } else {
                   q.addLast(curDoc);
               }
           }
        }

        System.out.println(sb);
        br.close();
    }

    static class Documents {
        int priority;
        int idx;

        public Documents(int priority, int idx) {
            this.priority = priority;
            this.idx = idx;
        }
    }
}